# Exercise 20
**Cristina Correa - 1819211867**


Now we will use multiple features to predict the grade and
compare the results to the single viable. To do so we need to redefine the X variable as follows.
(Note: we cannot plot multidimensional plots, thus you should not plot the data as in exercise
16)

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

file_name = "student-mat.csv"
df = pd.read_csv(file_name, sep=";")

numeric_df = df.select_dtypes(include=["float64", "int64"])

X = df[["G1", "G2", "age", "traveltime", "studytime", "failures"]]
y = df["G3"]  

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("\nResults for multiple features model:")
print(f"Mean Absolute Error: {mae:.4f}")
print(f"Mean Squared Error: {mse:.4f}")
print(f"R-squared: {r2:.4f}")

correlation_matrix = numeric_df.corr()
print("\nCorrelation matrix with G3:")
print(correlation_matrix["G3"].sort_values(ascending=False))


Results for multiple features model:
Mean Absolute Error: 1.3795
Mean Squared Error: 6.0690
R-squared: 0.7827

Correlation matrix with G3:
G3            1.000000
G2            0.904868
G1            0.801468
Medu          0.217147
Fedu          0.152457
studytime     0.097820
famrel        0.051363
absences      0.034247
freetime      0.011307
Walc         -0.051939
Dalc         -0.054660
health       -0.061335
traveltime   -0.117142
goout        -0.132791
age          -0.161579
failures     -0.360415
Name: G3, dtype: float64


**Look back at the correlation matrix and find the most useful features you may use to get the best results.**

- **G2** (0.9049): This has the highest positive correlation with G3. Since G2 represents the second period grade, it is a strong predictor of the final grade. Including G2 in the model is crucial.
- **G1** (0.8015): The first period grade (G1) is also highly correlated with G3 and should be included in the model.
- **Medu** (0.2171): The mother's education level is moderately correlated with G3 and could provide additional value in predicting student performance.
- **Fedu** (0.1525): The father's education level, while not as strongly correlated as G1 and G2, could still contribute valuable information.